lnrpc/walletrpc/psbt.go

```diff
@@ -0,0 +1,73 @@
+// +build walletrpc
+
+package walletrpc
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil/psbt"
+	"github.com/btcsuite/btcwallet/wtxmgr"
+	"github.com/lightningnetwork/lnd/lnwallet"
+)
+
+const (
+	defaultMinConf = 1
+	defaultMaxConf = math.MaxInt32
+)
+
+// utxoLock is a type that contains an outpoint of an UTXO and its lock lease
+// information.
+type utxoLock struct {
+	lockID     wtxmgr.LockID
+	outpoint   wire.OutPoint
+	expiration time.Time
+}
+
+// verifyInputsUnspent checks that all inputs are contained in the list of
+// known, non-locked UTXOs given.
+func verifyInputsUnspent(inputs []*wire.TxIn, utxos []*lnwallet.Utxo) error {
+	for idx, txIn := range inputs {
+		found := false
+		for _, u := range utxos {
+			if u.OutPoint == txIn.PreviousOutPoint {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return fmt.Errorf("input %d not found in list of non-"+
+				"locked UTXO", idx)
+		}
+	}
+
+	return nil
+}
+
+// lockInputs requests a lock lease for all inputs specified in a PSBT packet
+// by using the internal, static lock ID of lnd's wallet.
+func lockInputs(w lnwallet.WalletController, packet *psbt.Packet) ([]*utxoLock,
+	error) {
+
+	locks := make([]*utxoLock, len(packet.UnsignedTx.TxIn))
+	for idx, rawInput := range packet.UnsignedTx.TxIn {
+		lock := &utxoLock{
+			lockID:   LndInternalLockID,
+			outpoint: rawInput.PreviousOutPoint,
+		}
+
+		expiration, err := w.LeaseOutput(lock.lockID, lock.outpoint)
+		if err != nil {
+			return nil, fmt.Errorf("could not lease a lock on "+
```

Comment:
There should be `defer` to unlock the set of inputs that were already leased up until this point. Unless we want the caller to have to handle this responsibility.

Reply:
Right, good point! Decided to not go with a `defer` as we only have one error path. So coded it out explicitly there.

Reply:
What @Roasbeef was referring to was to defer at the level of the caller of `lockInputs`, as an error can occur in some other method after the fact and the inputs wouldn't be released. There isn't a method that's called after `lockInputs` that returns an error at the moment, but it's probably still worth making the change to prevent any potential bugs in the future.

Reply:
Ah, I understand now. But the `lockInputs()` is the last thing we do in `FundPsbt()` so nothing can fail after that. But I extended the comment on the function call and also on the RPC description that it's the caller's responsibility to unlock the UTXOs in case of an error on the caller's side.

I don't think we can make this nicer without introducing state to the walletkit server that keeps track of all PSBTs.

Reply:
another option would be to make a `LeaseOutputs` and `ReleaseOutputs` that lock all of them atomically

Reply:
Yeah, I agree, that would have been the best option. But that would require another `btcwallet` change which I'd rather avoid at this point to move forward with this feature.

Reply:
@guggero ok I think that's reasonable. Worst case someone doesn't have access to all their inputs until the lease expires.